### ISSUE TITLE
Training and validation accuracy very low in Binary Classification

### ISSUE THREAD

**Issue by Techro-Raj** [dmlc/mxnet#5245]
I'm trying to do binary classification to predict if a person is likely to vote (1) or not (0) based on 50 numeric features. I have 1261382 training examples, and I'm unsure about the batch size and optimal number of epochs I should have. For some reason, my training and validation accuracy is very low - around 0.00063. 

Here is my code:
```python
import mxnet as mx
import logging

logger = logging.getLogger()
logger.setLevel(logging.DEBUG)

batch_size = 612

train_iter = mx.io.CSVIter(data_csv="feature_train_v1.csv", data_shape=(50, ),
                   label_csv="label_train_v1.csv", label_shape= (1,), batch_size=batch_size,)

val_iter = mx.io.CSVIter(data_csv = "feature_validate_v1.csv", data_shape=(50, ),
                    label_csv="label_validate_v1.csv", label_shape= (1,), batch_size = batch_size)

net = mx.sym.Variable('data')
h1 = mx.sym.FullyConnected(data=net, name='h1', num_hidden=50)
net = mx.sym.BatchNorm(net, fix_gamma=True)
net = mx.sym.Activation(data=h1, act_type="sigmoid")
net = mx.symbol.Dropout(net)
h2 = mx.sym.FullyConnected(data=net, name = 'h2', num_hidden=1)
softmax = mx.sym.LogisticRegressionOutput(data=h2, name='softmax') # Calculate the Loss !

model = mx.mod.Module(
    softmax,
    data_names = ['data']
)

metric = mx.metric.CrossEntropy()

# train the network
model.fit(train_iter, eval_data=val_iter,
            optimizer_params={'learning_rate':0.1, 'momentum': 0.9},
            num_epoch=100)

model.score(val_iter, metric= metric)

# Predict:

#test_iter = mx.io.CSVIter(data_csv = "feature_test_v1.csv", data_shape=(1,50), batch_size = batch_size)
#model.predict(test_iter).asnumpy()
```

Any idea what I'm doing wrong or how I can improve the network? Thanks!

**Comment by Garglesoap** [comment]
What are you dropping out here?
    net = mx.symbol.Dropout(net)
I haven't seen that yet but I've just been covering the tutorials. Also maybe try lowering learning rate or momentum?

**Comment by 8Gitbrix** [comment]
@Garglesoap  When you call dropout on net, at each training epoch, it drops out random neurons in the layer with a probability of p (0.5 in my example). The first dropout drops out random neurons in the first layer (h1) in the example below, and the second one drops out random neurons in layer h2. By dropping random neurons, dropout prevents overfitting because it forces the network to learn combinations of more robust features instead of only fine tuning weights of features that may be insignificant in a normal neural network. As a result, it might / probably will take longer to train the network. 

Here is my fixed network:
```python
from sklearn.metrics import roc_auc_score, auc
import mxnet as mx

batch_size = 10000
train_iter = mx.io.CSVIter(data_csv="feature_train_v1.csv", data_shape=(50, ),
                   label_csv="label_train_v1.csv", label_shape= (1,), batch_size=batch_size,)

val_iter = mx.io.CSVIter(data_csv = "feature_validate_v1.csv", data_shape=(50, ),
                    label_csv="label_validate_v1.csv", label_shape= (1,), batch_size = batch_size)

data = mx.sym.Variable('data')
net = mx.sym.BatchNorm(data, fix_gamma=True)
net = mx.symbol.Dropout(net, p=0.5)
h1 = mx.sym.FullyConnected(data=net, name='h1', num_hidden=50)
net = mx.sym.Activation(data=h1, act_type="relu")

net = mx.sym.BatchNorm(net, fix_gamma=True)
net = mx.symbol.Dropout(net, p=0.5)
h2 = mx.sym.FullyConnected(data=net, name='h2', num_hidden=25)
net = mx.sym.Activation(data=h2, act_type="relu")

net = mx.sym.BatchNorm(net, fix_gamma=True)
h3 = mx.sym.FullyConnected(data=net, name ='h3', num_hidden=1)

softmax = mx.sym.LogisticRegressionOutput(data=h3, name='softmax') # Calculate the Loss !

model = mx.mod.Module(softmax)

# train the network
model.fit( train_iter,
            eval_data=val_iter,
            eval_metric= roc_auc_score,
            optimizer_params={'learning_rate':0.01, 'momentum': 0.9},
            num_epoch=9)

model.score(val_iter, roc_auc_score)
```

I fixed my error after realizing a few issues. 
* Since it is a classification problem, it helps to use metrics like AUC or cross entropy.
* Try using at least 3 fully connected layers. 
If you are doing a classification problem and are using sklearn's AUC function, it won't work for datasets that don't have enough negative labels (in a binary classification function). 

I found MXNet's website and tutorial a bit outdated and hard to follow when I was learning it around January. I had to experiment by reading the github repo - you'd probably have to do the same :| . 

I no longer use MXNet (I started using pytorch :) ... its nice!), but hope this helps.

**Comment by wythe51** [comment]
Maybe try `SoftmaxOutput` instead of `LogisticRegressionOutput` (and set `num_hidden=2` ).
At least in my case, it works.
